proposals/NNNN-typed-throws-concurrency-library.md

```diff
@@ -0,0 +1,319 @@
+) -> Task<Success, Failure>
+```
+
+Finally, the `value` property of `Task` is similarly overloaded:
+
+```swift
+extension Task where Failure == Never {}
+  var value: Success { get async }
+}
+extension Task where Failure == any Error {
+  var value: Success { get async throws }
+}
+```
+
+These two can be replaced with a single property:
+
+```swift
+var value: Success { get async throws(Failure) }
+```
+
+### Continuations
+
+Currently, there are two variants of each continuation creation method to
+accommodate for the throwing and non-throwing variant. Since, the natural
+spelling would be without the `Throwing` in the method name. We propose
+to add the following two new methods:
+
+```swift
+public func withCheckedContinuation<T, Failure: Error>(
+    function: String = #function,
+    _ body: (CheckedContinuation<T, Failure>) -> Void
+) async throws(Failure) -> T
+
+public func withUnsafeContinuation<T, Failure: Error>(
+  _ fn: (UnsafeContinuation<T, Failure>) -> Void
+) async throws(Failure) -> T
+```
+
+#### Task cancellation
+
+A few of the `Task` APIs are documented to only throw `CancellationError` and
+can adopt typed throws. For example, `checkCancellation`:
+
+```swift
+public static func checkCancellation() throws(CancellationError)
+```
+
+Similarly, the `sleep` APIs will only throw on cancellation:
+
+```swift
+public static func sleep<C: Clock>(
+  until deadline: C.Instant,
+  tolerance: C.Instant.Duration? = nil,
+  clock: C = ContinuousClock()
+) async throws(CancellationError)
+
+public static func sleep<C: Clock>(
+  for duration: C.Instant.Duration,
+  tolerance: C.Instant.Duration? = nil,
+  clock: C = ContinuousClock()
+) async throws(CancellationError)
+```
+
+Additionally, the `ContinuousClock` and `SuspendingClock` `sleep` methods will also
+adopt typed throws:
+
+```swift
+public func sleep(
+  until deadline: Instant, tolerance: Swift.Duration? = nil
+) async throws(CancellationError)
+```
+
+Lastly, the `withTaskCancellationHandler` needs a new variant
+that supports a typed error `operation` closure:
+
+```swift
+public func withTaskCancellationHandler<T, Failure: Error>(
+  operation: () async throws(Failure) -> T,
+  onCancel handler: @Sendable () -> Void
+) async rethrows -> T
+```
+
+### Task groups
+
+The task group APIs also need to gain support for typed throws. The important
+detail here is that there are two layers of failures involved. First, the
+failures produced by the child task themselves. Secondly, the failure returned
+by `withThrowingTaskGroup` API itself. We propose to add the following API:
+
+```swift
+public func withTaskGroup<ChildTaskResult, ChildTaskFailure: Error, GroupResult, GroupFailure: Error>(
+  of childTaskResultType: ChildTaskResult.Type,
+  childTaskFailureType: ChildTaskFailure.Type,
+  returning returnType: GroupResult.Type = GroupResult.self,
+  body: (inout ThrowingTaskGroup<ChildTaskResult, ChildTaskFailure>) async throws(GroupFailure) -> GroupResult
```

Comment:
I think it would be good if the body could throw a different type independently of the group e.g.:
```swift
public func withTaskGroup<ChildTaskResult, ChildTaskFailure: Error, GroupResult, GroupFailure: Error, BodyFailure: Error>(
    of childTaskResultType: ChildTaskResult.Type,
    childTaskFailureType: ChildTaskFailure.Type,
    returning returnType: GroupResult.Type = GroupResult.self,
    bodyFailureType: BodyFailure.self,
    body: (inout ThrowingTaskGroup<ChildTaskResult, ChildTaskFailure>) async throws(GroupFailure) -> GroupResult
) async throws(BodyFailure) -> GroupResult
```
This would also allow to throw an error even if the child task can't throw an error because you do e.g. decoding of the result of all child tasks after all child tasks have finished.

Reply:
The proposed API already allows that. There is a separate `ChildTaskFailure` and `GroupFailure` type

Reply:
Yeah this looks ok; though worried that spelling out `childTaskFailureType` is quite a lot... but i guess there's difficulty in finding a good name here ("childTaskThrowing: " ?)